File: scripts/translation_service.py
"""Reliable Turkish-to-locale translation for generated landing content.

Gemini is preferred when GEMINI_API_KEY is available. The anonymous GTX endpoint
is retained as a fallback. A failed call returns None: callers must preserve an
existing page or stop the generation, never write Turkish into another locale.
"""

import json
import os
import time
import urllib.error
import urllib.parse
import urllib.request
import re

GEMINI_MODEL = os.environ.get("GEMINI_MODEL") or os.environ.get("TREESCOUT_TRANSLATION_MODEL") or "gemini-2.5-flash"
_GEMINI_PAUSED_UNTIL = 0.0


def _retry_delay(attempt: int) -> float:
    return min(2.0 * (2**attempt), 12.0)


def _http_retry_delay(error: urllib.error.HTTPError, attempt: int) -> float:
    """Honor provider Retry-After/retryDelay without exposing response bodies."""
    header = error.headers.get('Retry-After') if error.headers else None
    if header:
        try:
            return min(max(float(header), 1.0), 90.0)
        except ValueError:
            pass
    try:
        body = error.read().decode('utf-8', errors='replace')
        match = re.search(r'"retryDelay"\s*:\s*"([0-9.]+)s"', body)
        if match:
            return min(max(float(match.group(1)), 1.0), 90.0)
    except Exception:
        pass
    return _retry_delay(attempt)


def _pause_gemini(delay: float) -> None:
    global _GEMINI_PAUSED_UNTIL
    _GEMINI_PAUSED_UNTIL = max(_GEMINI_PAUSED_UNTIL, time.time() + delay)


def _gemini_is_paused() -> bool:
    return time.time() < _GEMINI_PAUSED_UNTIL


def _gemini(text: str, lang: str, key: str) -> str | None:
    if _gemini_is_paused():
        return None
    body = {
        "systemInstruction": {
            "parts": [{
                "text": (
                    "You are a precise professional translator. Translate Turkish into the requested language. "
                    "Return only the translation, with no quotation marks, commentary, markdown, or language label. "
                    "Do not summarize, omit, or add claims. Preserve product names, repository names, URLs and numbers."
                )
            }]
        },
        "contents": [{
            "parts": [{
                "text": (
                    f"Translate this Turkish technology-site text into {lang}. "
                    "Keep the meaning and register natural for the target locale.\n\n{text}"
                )
            }]
        }],
        "generationConfig": {
            "temperature": 0.1,
            "maxOutputTokens": 2048,
        },
    }
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{GEMINI_MODEL}:generateContent"
    request = urllib.request.Request(
        url,
        data=json.dumps(body).encode("utf-8"),
        headers={"Content-Type": "application/json", "x-goog-api-key": key},
        method="POST",
    )
    for attempt in range(4):
        try:
            with urllib.request.urlopen(request, timeout=60) as response:
                payload = json.loads(response.read().decode("utf-8"))
            candidates = payload.get("candidates") or []
            parts = (candidates[0].get("content") or {}).get("parts") or []
            result = "".join(str(part.get("text") or "") for part in parts).strip()
            if result:
                if result.startswith("```") and result.endswith("```"):
                    result = result.split("\n", 1)[-1][:-3].strip()
                return result or None
            raise ValueError("Gemini empty translation response")
        except urllib.error.HTTPError as error:
            if error.code == 429:
                _pause_gemini(_http_retry_delay(error, attempt))
                return None
            if error.code not in (500, 502, 503) or attempt == 3:
                return None
            time.sleep(_http_retry_delay(error, attempt))
            continue
        except Exception:
            if attempt == 3:
                return None
        time.sleep(_retry_delay(attempt))
    return None


def _gtx(text: str, lang: str) -> str | None:
    url = (
        "https://translate.googleapis.com/translate_a/single?client=gtx&sl=tr&"
        f"tl={urllib.parse.quote(lang)}&dt=t&q={urllib.parse.quote(text)}"
    )
    request = urllib.request.Request(url, headers={"Accept": "application/json", "User-Agent": "TreScout/1.0"})
    for attempt in range(3):
        try:
            with urllib.request.urlopen(request, timeout=20) as response:
                payload = json.loads(response.read().decode("utf-8"))
            if not isinstance(payload, list) or not payload or not isinstance(payload[0], list):
                return None
            result = "".join(str(part[0]) for part in payload[0] if isinstance(part, list) and part).strip()
            return result or None
        except urllib.error.HTTPError as error:
            if error.code == 429:
                _pause_gemini(_http_retry_delay(error, attempt))
                return None
            if error.code not in (429, 500, 502, 503) or attempt == 2:
                return None
            time.sleep(_http_retry_delay(error, attempt))
            continue
        except Exception:
            if attempt == 2:
                return None
        time.sleep(_retry_delay(attempt))
    return None


def translate_text(text: str, lang: str) -> str | None:
    clean = (text or "").strip()
    if not clean:
        return ""
    key = os.environ.get("GEMINI_API_KEY", "").strip()
    if key:
        translated = _gemini(clean, lang, key)
        if translated:
            return translated
    return _gtx(clean, lang)


def _gemini_batch(texts: list[str], lang: str, key: str) -> dict[str, str] | None:
    items = [{"id": str(index), "text": text} for index, text in enumerate(texts)]
    body = {
        "systemInstruction": {
            "parts": [{
                "text": (
                    "You are a precise professional translator. Output valid JSON only. Translate each Turkish "
                    "item into the requested language. Preserve every id exactly, do not summarize, omit, merge, "
                    "or add items, and preserve proper nouns, URLs, numbers, and technical meaning."
                )
            }]
        },
        "contents": [{
            "parts": [{
                "text": (
                    f"Translate every item into {lang}. Return a JSON array with exactly one object per input, "
                    "using the same id and the translated text in the text field.\n\n" + json.dumps(items, ensure_ascii=False)
                )
            }]
        }],
        "generationConfig": {
            "temperature": 0.1,
            "maxOutputTokens": 8192,
            "responseMimeType": "application/json",
        },
    }
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{GEMINI_MODEL}:generateContent"
    request = urllib.request.Request(
        url,
        data=json.dumps(body, ensure_ascii=False).encode("utf-8"),
        headers={"Content-Type": "application/json", "x-goog-api-key": key},
        method="POST",
    )
    for attempt in range(3):
        try:
            with urllib.request.urlopen(request, timeout=90) as response:
                payload = json.loads(response.read().decode("utf-8"))
            candidates = payload.get("candidates") or []
            parts = (candidates[0].get("content") or {}).get("parts") or []
            raw = "".join(str(part.get("text") or "") for part in parts).strip()
            if raw.startswith("```"):
                raw = raw.split("\n", 1)[-1].removesuffix("```").strip()
            parsed = json.loads(raw)
            rows = parsed if isinstance(parsed, list) else parsed.get("translations")
            if not isinstance(rows, list) or len(rows) != len(items):
                raise ValueError("Gemini batch translation shape mismatch")
            result: dict[str, str] = {}
            for row in rows:
                if not isinstance(row, dict):
                    raise ValueError("Gemini batch row is not an object")
                index = int(str(row.get("id", "")))
                if index < 0 or index >= len(items):
                    raise ValueError("Gemini batch id is invalid")
                value = str(row.get("text") or "").strip()
                if not value:
                    raise ValueError("Gemini batch item is empty")
                result[items[index]["text"]] = value
            if len(result) != len(items):
                raise ValueError("Gemini batch ids are not unique")
            return result
        except urllib.error.HTTPError as error:
            if error.code == 429:
                _pause_gemini(_http_retry_delay(error, attempt))
                return None
            if error.code not in (429, 500, 502, 503) or attempt == 2:
                return None
            time.sleep(_http_retry_delay(error, attempt))
            continue
        except Exception:
            if attempt == 2:
                return None
        time.sleep(_retry_delay(attempt))
    return None


def translate_texts(texts: list[str], lang: str) -> dict[str, str | None]:
    """Translate passages in paced batches; never return source text on failure."""
    unique = list(dict.fromkeys((text or "").strip() for text in texts if (text or "").strip()))
    result: dict[str, str | None] = {}
    key = os.environ.get("GEMINI_API_KEY", "").strip()
    batch_size = 12
    for start in range(0, len(unique), batch_size):
        batch = unique[start:start + batch_size]
        translated = _gemini_batch(batch, lang, key) if key else None
        if translated is None:
            # Do not immediately fire Gemini once per item after a batch quota
            # failure. GTX is bounded and failures remain explicit as None.
            translated = {}
            for text in batch:
                value = _gtx(text, lang)
                if value:
                    translated[text] = value
        for text in batch:
            result[text] = translated.get(text)
        if start + batch_size < len(unique):
            time.sleep(5.0)
    return result

File: scripts/translation-service.js
const https = require('https');

const MODEL = process.env.GEMINI_MODEL || process.env.TREESCOUT_TRANSLATION_MODEL || 'gemini-2.5-flash';

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

function requestJson(url, options, body, timeoutMs) {
  return new Promise((resolve, reject) => {
    const req = https.request(url, { ...options, timeout: timeoutMs }, res => {
      let raw = '';
      res.setEncoding('utf8');
      res.on('data', chunk => { raw += chunk; });
      res.on('end', () => {
        let data;
        try { data = JSON.parse(raw); } catch { reject(new Error(`invalid JSON HTTP ${res.statusCode}`)); return; }
        if ((res.statusCode || 500) < 200 || (res.statusCode || 500) >= 300) {
          reject(new Error(`HTTP ${res.statusCode}: ${JSON.stringify(data).slice(0, 180)}`));
          return;
        }
        resolve(data);
      });
    });
    req.on('timeout', () => req.destroy(new Error('translation request timeout')));
    req.on('error', reject);
    req.write(body);
    req.end();
  });
}

function clean(value) {
  return String(value || '')
    .trim()
    .replace(/^```(?:text|plaintext)?\s*/i, '')
    .replace(/\s*```$/, '')
    .trim();
}

async function gemini(text, lang) {
  const key = (process.env.GEMINI_API_KEY || '').trim();
  if (!key) return null;
  const body = JSON.stringify({
    systemInstruction: { parts: [{ text:
      'You are a precise professional translator. Translate Turkish into the requested language. ' +
      'Return only the translation, without quotation marks, commentary, markdown, or language labels. ' +
      'Do not summarize, omit, or add claims. Preserve product names, repository names, URLs, and numbers.' }] },
    contents: [{ parts: [{ text:
      `Translate this Turkish technology-site text into ${lang}. Keep the meaning natural for the target locale.\n\n${text}` }] }],
    generationConfig: { temperature: 0.1, maxOutputTokens: 2048 },
  });
  const url = `https://generativelanguage.googleapis.com/v1beta/models/${MODEL}:generateContent`;
  for (let attempt = 0; attempt < 4; attempt += 1) {
    try {
      const data = await requestJson(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', 'x-goog-api-key': key },
      }, body, 60000);
      const textParts = data?.candidates?.[0]?.content?.parts || [];
      const result = clean(textParts.map(part => part?.text || '').join(''));
      if (result) return result;
      throw new Error('Gemini empty translation response');
    } catch (error) {
      const msg = String(error?.message || error);
      const retryable = /HTTP (429|500|502|503)|timeout|ECONNRESET|UNAVAILABLE|RESOURCE_EXHAUSTED/i.test(msg);
      if (!retryable || attempt === 3) return null;
      await sleep(Math.min(1200 * (2 ** attempt), 12000));
    }
  }
  return null;
}

async function gtx(text, lang) {
  const url = `https://translate.googleapis.com/translate_a/single?client=gtx&sl=tr&tl=${encodeURIComponent(lang)}&dt=t&q=${encodeURIComponent(text)}`;
  for (let attempt = 0; attempt < 3; attempt += 1) {
    try {
      const data = await requestJson(url, { method: 'GET', headers: { Accept: 'application/json', 'User-Agent': 'TreScout/1.0' } }, '', 20000);
      const result = clean((data?.[0] || []).map(part => part?.[0] || '').join(''));
      if (result) return result;
      throw new Error('GTX empty translation response');
    } catch (error) {
      const msg = String(error?.message || error);
      const retryable = /HTTP (429|500|502|503)|timeout|ECONNRESET/i.test(msg);
      if (!retryable || attempt === 2) return null;
      await sleep(Math.min(1500 * (2 ** attempt), 8000));
    }
  }
  return null;
}

async function translateText(text, lang) {
  const cleanText = String(text || '').trim();
  if (!cleanText) return '';
  return (await gemini(cleanText, lang)) || (await gtx(cleanText, lang));
}

module.exports = { translateText };


function parseJsonPayload(raw) {
  const trimmed = String(raw || '').trim();
  try {
    return JSON.parse(trimmed);
  } catch {
    const fenceMatch = trimmed.match(/```(?:json)?\s*([\s\S]*?)\s*```/i);
    if (fenceMatch) {
      try {
        return JSON.parse(fenceMatch[1].trim());
      } catch {}
    }
    const arrayMatch = trimmed.match(/\[\s*\{[\s\S]*\}\s*\]/);
    if (arrayMatch) {
      try {
        return JSON.parse(arrayMatch[0].trim());
      } catch {}
    }
    throw new Error('Could not parse JSON from Gemini response');
  }
}

async function geminiBatch(texts, lang) {
  const key = (process.env.GEMINI_API_KEY || '').trim();
  if (!key || !texts.length) return null;
  const body = JSON.stringify({
    systemInstruction: { parts: [{ text:
      'You are a precise professional translator. Output valid JSON only. Translate each Turkish item into the requested language. ' +
      'Preserve every id exactly, do not summarize, omit, merge, or add items. Preserve proper nouns, URLs, numbers, and technical meaning.' }] },
    contents: [{ parts: [{ text:
      `Translate every item into ${lang}. Return a JSON array with exactly one object per input, using the same id and the translated text in the text field.\n\n` +
      JSON.stringify(texts.map((text, index) => ({ id: String(index), text }))) }] }],
    generationConfig: { temperature: 0.1, maxOutputTokens: 8192, responseMimeType: 'application/json' },
  });
  const url = `https://generativelanguage.googleapis.com/v1beta/models/${MODEL}:generateContent`;
  for (let attempt = 0; attempt < 3; attempt += 1) {
    try {
      const data = await requestJson(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', 'x-goog-api-key': key },
      }, body, 90000);
      const raw = clean((data?.candidates?.[0]?.content?.parts || []).map(part => part?.text || '').join(''));
      const parsed = parseJsonPayload(raw);
      const rows = Array.isArray(parsed) ? parsed : parsed?.translations;
      if (!Array.isArray(rows) || rows.length !== texts.length) throw new Error('Gemini batch shape mismatch');
      const result = new Map();
      for (const row of rows) {
        const index = Number(row?.id);
        const value = clean(row?.text);
        if (!Number.isInteger(index) || index < 0 || index >= texts.length || !value) throw new Error('Gemini batch row invalid');
        result.set(texts[index], value);
      }
      if (result.size !== texts.length) throw new Error('Gemini batch ids are not unique');
      return result;
    } catch (error) {
      const msg = String(error?.message || error);
      const retryable = /HTTP (429|500|502|503)|timeout|ECONNRESET|UNAVAILABLE|RESOURCE_EXHAUSTED|shape mismatch|Could not parse JSON|SyntaxError/i.test(msg);
      if (!retryable || attempt === 2) return null;
      await sleep(Math.min(1500 * (2 ** attempt), 12000));
    }
  }
  return null;
}

async function translateTexts(texts, lang) {
  const unique = [...new Set(texts.map(text => String(text || '').trim()).filter(Boolean))];
  const result = new Map();
  const batchSize = 12;
  for (let start = 0; start < unique.length; start += batchSize) {
    const batch = unique.slice(start, start + batchSize);
    const translated = await geminiBatch(batch, lang);
    if (translated) {
      for (const [source, value] of translated) result.set(source, value);
    } else {
      // If batch fails, attempt single Gemini calls before falling back to GTX
      for (const source of batch) {
        const singleGemini = await gemini(source, lang);
        if (singleGemini) {
          result.set(source, singleGemini);
        } else {
          const gtxVal = await gtx(source, lang);
          if (gtxVal) result.set(source, gtxVal);
        }
      }
    }
    if (start + batchSize < unique.length) await sleep(3000);
  }
  return result;
}

module.exports = { translateText, translateTexts };
